### PR TITLE
You can now catch thrown items by having your throw intent on and an empty active hand

### PR DESCRIPTION
**Pull request by 9600bauds**
technikally a port from TG

Hilariously shitty demonstration:

You cannot catch items unless you have throw intent on. You must have your ACTIVE hand selected to catch something (you can't have your esword in your active hand and keep throw intent on with your other hand empty). Since your throwing intent gets reset when you catch something you need to free your hand and then re-enable it after every catch.

Will conflict with #13146 since I'm using the same new define file that he created there, but it would be trivial for either of us to fix that conflict I think. His PR should be merged before mine anyways.

:cl:
 * rscadd: You can now catch items that were thrown at you by having throw intent enabled and having your active hand empty.

### CODE REVIEW

**Comment by Sprok0** [comment]
Time to catch pneumatic cannon shards

**Comment by 9600bauds** [comment]
Can't catch things going 2fast, I made sure of that ',:^)

**Comment by Sprok0** [comment]
aww

**Comment by ComicIronic** [comment]
>multiclient

Ban this sick filth.

**Comment by Probe1** [comment]
>Can't catch things going 2fast, I made sure of that ',:^)

the dream of bullet catching is dead

**Comment by AI-Is-Rogue** [comment]
Can you catch things thrown at you from behind?

File: code/modules/mob/living/carbon/combat.dm [hunk 1]
```diff
@@ -3,3 +3,12 @@
 		share_contact_diseases(C)
 
 	return ..()
+
+/mob/living/carbon/hitby(var/obj/item/I, var/speed, var/dir)
+	if(istype(I) && isturf(I.loc))
+		if(!restrained() && in_throw_mode && !get_active_hand()) //We're an able-bodied person with an empty hand and intent to catch
+			if(speed < EMBED_THROWING_SPEED && put_in_hands(I)) //Can't catch things going too fast
+				visible_message("<span class='warning'>[src] catches \the [I]!</span>")
```
Comment (PJB3005) [inline code/modules/mob/living/carbon/combat.dm]: Put `\The [src]`

File: __DEFINES/game.dm [hunk 2]
```diff
@@ -0,0 +1 @@
+#define EMBED_THROWING_SPEED 20
```
Comment (PJB3005) [inline __DEFINES/game.dm:1]: What kind of generic filename is this?

**Comment by Probe1** [comment]
@AI-is-rogue is right.  Make sure it checks ```dir```.

**Comment by kol1th** [comment]
Genetic mutation that lets you catch projectiles and throw them back when?

**Comment by 9600bauds** [comment]
I don't think there's anything else that checks which way you're facing, even the PR that made shields work like that was closed

The thing with spess movement is that you don't manually turn to face things, you auto-face anything you shoot or move towards. And until you can run backwards so you can go long...

There's also the problem that for some reason thrown items are several times faster than any bullet so I can't expect anyone to have the reaction time to do something this insignificant.

File: __DEFINES/game.dm (hunk 2 above)
Comment (9600bauds) [reply __DEFINES/game.dm:1]: c OP, it will be aded in #13146

Comment (PJB3005) [reply __DEFINES/game.dm:1]: Goddamnit the one time I don't review a PR and shit like this gets through.

Comment (9600bauds) [reply __DEFINES/game.dm:1]: There IS a distinct lack of an appropriately named _DEFINE to put shit like this in, though

**Comment by Probe1** [comment]
its a feaaaaature

**Comment by Exxion** [comment]
You *can* run backwards if you know how

**Comment by Kammerjunk** [comment]
I thought this existed when I started playing and for a long while since. I _swear_ it said so on some wiki page.
👍

**Comment by Exxion** [comment]
Also SCP-173 does check which way you're facing
But that's the only thing I know of

**Comment by ComicIronic** [comment]
Be sure to do a `dir & src.dir` to allow for diagonal catching.

**Comment by pvak** [comment]
can you catch fireballs?

**Comment by Exxion** [comment]
:|

**Comment by Exxion** [comment]
@ComicIronic :|

@pvak Fireballs are projectiles, not thrown items.

**Comment by 9600bauds** [comment]
No you cannot catch projectiles
Hey look you can do sick trickcatches

**Comment by ComicIronic** [comment]
What, you don't want to be able to catch diagonally?

**Comment by Exxion** [comment]
Try reading the comment thread more carefully.

**Comment by ComicIronic** [comment]
Spoonfeed me you obtuse shit.

**Comment by Exxion** [comment]
It doesn't check `dir` at all and Bauds doesn't intend to make it do so

**Comment by ComicIronic** [comment]
Without a dir check, I don't see how this doesn't make disarm the best intent but better.

**Comment by ComicIronic** [comment]
Without a dir check, wouldn't people just throw intent defensively in a  chase?

**Comment by 9600bauds** [comment]
People can't even set walk intent to go through the wet floor they've already slipped twice on

**Comment by 9600bauds** [comment]
This is also apparently how it works on TG, I asked around because I thought it was a neat idea and even with the possibility of defensive catching nobody can remember it exists to use it effectively

File: __DEFINES/game.dm (hunk 2 above)
Comment (PJB3005) [reply __DEFINES/game.dm:1]: Then make one damnit.

File: code/modules/mob/living/carbon/combat.dm [hunk 3]
```diff
@@ -3,3 +3,12 @@
...
 	return ..()
+
+/mob/living/carbon/hitby(var/obj/item/I, var/speed, var/dir)
+	if(istype(I) && isturf(I.loc))
+		if(!restrained() && in_throw_mode && !get_active_hand()) //We're an able-bodied person with an empty hand and intent to catch
+			if(speed < EMBED_THROWING_SPEED && put_in_hands(I)) //Can't catch things going too fast
+				visible_message("<span class='warning'>\The [src] catches \the [I]!</span>")
+				throw_mode_off()
+				return 1
+	..()
```
Comment (ComicIronic) [inline code/modules/mob/living/carbon/combat.dm]: It looks like `hitby` has a return value, so return this.

Comment (9600bauds) [reply code/modules/mob/living/carbon/combat.dm]: 100% correct thank you Comic

**Comment by Probe1** [comment]
how often do people throw shit at each other in a chase

i've never actually seen bolas used

**Comment by SonixApache** [comment]
Flashbangs, maybe?

**Comment by 9600bauds** [comment]
Bolas are actually shit because even the quality bolas have a 40% chance to not work (ghetto bolas only having a 30% chance to work at all)
I plan to remove that sometime

**Comment by Shadowmech88** [comment]
>Can't catch things going 2fast

What about making it so that you can if you're hulked/strong mutation/some other special condition?
Like a ninja catching an arrow mid-flight.

**Comment by Sprok0** [comment]
As long as thrown items are literally faster than bullets I don't think expecting people to face incoming projectiles is reasonable

**Comment by 9600bauds** [comment]
Isn't there a "fast" Genetics mutation that does essentially nothing?

**Comment by Shadowmech88** [comment]
I'm pretty sure the fast mutation *used* to make you actually faster, but ever since whatever movecode change that negated all sources of positive speed like hyperzine and the vtec module, it just removes slowdown.

It could always be fixed, though.

File: __DEFINES/game.dm (hunk 2 above)
Comment (9600bauds) [reply __DEFINES/game.dm:1]: But PJB

You are commenting in it

**Review by ComicIronic — APPROVED** [review]

**Comment by 9600bauds** [comment]
I legitimately can't find the mutation at all. Is it still in the code?

EDIT: nevermind, the wiki is just outdated

**Comment by 9600bauds** [comment]
There, because everything in this game needs a sekrit use that will probably only matter twice per year, the noslowdown Genetics mutation and Methylin now let you catch pneumatic cannon items.

Also I guess a proc that could be overwritten later for... ninjas or something.

**Comment by Kammerjunk** [comment]
If you have throw intent on and are holding floor tiles and someone throws a floor tile at you
Does it stack?

**Comment by 9600bauds** [comment]
No, you need to have your active hand empty
At that point the single floortile goes to your then-empty hand

**Comment by TheFriggenSawbones** [comment]
one of the scariest things ive ever seen, playing ss13, is some dweeb on fallout13 catching everything i threw at him using macros.

If that can't happen this pr is gucci.

**Comment by Exxion** [comment]
You can almost certainly do it with macros
But thrown items here are pretty wimpy anyway, unless they're thrown at extreme speed, and you can't catch past a certain speed

**Comment by 9600bauds** [comment]
If by macros you mean "slam your fat pudgy cheeto digit in such a way to press the drop item and throw intent keys at the same time"

**Comment by Exxion** [comment]
I forgot that throw is only a toggle, so while you COULD do it with a macro, it wouldn't really be any easier than doing it by hand

**Pull request merged**